geomesa-cassandra/geomesa-cassandra-datastore/src/main/scala/org/locationtech/geomesa/cassandra/index/CassandraFeatureIndex.scala

```diff
@@ -0,0 +1,186 @@
+/***********************************************************************
+* Copyright (c) 2013-2016 IBM
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Apache License, Version 2.0
+* which accompanies this distribution and is available at
+* http://www.opensource.org/licenses/apache2.0.php.
+*************************************************************************/
+
+package org.locationtech.geomesa.cassandra.index
+
+import com.google.common.primitives.{Bytes, Longs, Ints, Shorts}
+
+import com.datastax.driver.core._
+import com.datastax.driver.core.querybuilder._
+import com.vividsolutions.jts.geom.Geometry
+import org.geotools.factory.Hints
+import org.locationtech.geomesa.cassandra._
+import org.locationtech.geomesa.cassandra.data._
+import org.locationtech.geomesa.curve.BinnedTime
+import org.locationtech.geomesa.index.api.FilterStrategy
+import org.locationtech.geomesa.index.index.ClientSideFiltering.RowAndValue
+import org.locationtech.geomesa.index.index.{ClientSideFiltering, IndexAdapter}
+import org.locationtech.geomesa.curve.TimePeriod
+import org.opengis.feature.simple.SimpleFeatureType
+import org.opengis.feature.simple.SimpleFeature
+import org.opengis.filter.Filter
+
+import scala.collection.JavaConversions._
+
+import java.nio.ByteBuffer
+import java.nio.charset.StandardCharsets
+import java.util.Date
+
+object CassandraFeatureIndex extends CassandraIndexManagerType {
+  // note: keep in priority order for running full table scans
+  override val AllIndices: Seq[CassandraFeatureIndex] =
+    Seq(CassandraZ3Index, CassandraXZ3Index, CassandraZ2Index, CassandraXZ2Index, CassandraIdIndex, CassandraAttributeIndex)
+
+  override val CurrentIndices: Seq[CassandraFeatureIndex] = AllIndices
+}
+
+trait CassandraFeatureIndex extends CassandraFeatureIndexType
+    with IndexAdapter[CassandraDataStore, CassandraFeature, Statement, Statement] with ClientSideFiltering[Row] {
+  import org.locationtech.geomesa.utils.geotools.RichSimpleFeatureType.RichSimpleFeatureType
+
+  type RowToValues = (Array[Byte]) => Any
+
+  var getRowValues:RowToValues = {row => (0L, 0L, "")}
+  var tableName = ""
+  var ks = ""
+
+  def getRowValuesFromSFT(sft: SimpleFeatureType): RowToValues = {
+    // table sharing + shard + 2 byte short + 8 byte long + fid
+    val sharing = sft.getTableSharingPrefix
+    (row) => {
+      var curr = if (sharing == "") 0 else 1
+      // use shard and bin for pk
+      val pk:java.lang.Integer = ((row(curr) & 0xFF) << 16) | ((row(curr + 1) & 0xFF) << 8) | (row(curr + 2) & 0xFF);
+      curr += 3
+      val z = ByteBuffer.wrap(row, curr, 8).getLong
+      curr += 8
+      val fid = new String(row, curr, row.length - curr, StandardCharsets.UTF_8)
+      (pk, z, fid, sharing)
+    }
+  }
+
+  override def configure(sft: SimpleFeatureType, ds: CassandraDataStore): Unit = {
+    super.configure(sft, ds)
+    tableName = getTableName(sft.getTypeName, ds)
```

Comment:
this method will only get called when a new schema is created, so shouldn't be used to initialize state. also, table name is going to vary if you create different schemas, so shouldn't be stored in the common object. Instead you can pass it through in CassandraFeatureWriter.createMutators